### PR TITLE
type coercion: support is/is_not_`bool`/like/unknown expr

### LINKED ISSUE #3509: support the type coercion for `like` `unlike` `istrue` `isfalse` `isunknown`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

block this issue: https://github.com/apache/arrow-datafusion/issues/3388

- [ ] IsNotTrue IsTrue IsFalse  IsNotFalse
- [ ] IsUnknown IsNotUnknown
- [ ] Like,ILike
- [ ] between

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

Closes #3509 

block the pr https://github.com/apache/arrow-datafusion/pull/3396

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by liukun4515** [comment]
We will add the `like,ilike` and other expr in the type coercion optimizer rule.

I also has the question about that https://github.com/apache/arrow-datafusion/pull/3396#issuecomment-1248958557
https://github.com/apache/arrow-datafusion/pull/3396#issuecomment-1248962110

File: datafusion/expr/src/expr.rs
Comment (Dandandan) [inline datafusion/expr/src/expr.rs:500]:
```suggestion
    /// Return `IsTrue(Box(self))`
```

File: datafusion/expr/src/expr.rs
Comment (Dandandan) [inline datafusion/expr/src/expr.rs:505]:
```suggestion
    /// Return `IsNotTrue(Box(self))`
```

File: datafusion/expr/src/expr.rs
Comment (Dandandan) [inline datafusion/expr/src/expr.rs:510]:
```suggestion
    /// Return `IsFalse(Box(self))`
```

File: datafusion/expr/src/expr.rs
Comment (Dandandan) [inline datafusion/expr/src/expr.rs:515]:
```suggestion
    /// Return `IsNotFalse(Box(self))`
```

**Comment by liukun4515** [comment]
@alamb @andygrove @Dandandan PTAL

**Comment by andygrove** [comment]
Thanks @liukun4515 I will review this today or tomorrow

File: datafusion/optimizer/src/type_coercion.rs
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:139]: 🤔  maybe in the long term figuring out how to avoid having to remember to call the const evaluator each struct would be good -- as it too does a expr tree walk

So this algorithm is at least `O(n^2)` in the number of items in the expr tree. 

I don't think it is important to do in this PR, but I think it is worth keeping in mind as we proceed or if we see the planner being a bit slow

File: datafusion/optimizer/src/type_coercion.rs
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:166]: Is there a reason this isn't `Operator::IsUnknown`?

File: datafusion/optimizer/src/type_coercion.rs
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:175]: Likewise here, it seem strange that this is not `Operator::IsNotUnknown`

File: datafusion/optimizer/src/type_coercion.rs
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:300]:
```suggestion
// Support the `IsTrue` `IsNotTrue` `IsFalse` `IsNotFalse` type coercion.
```

**Review by alamb — APPROVED** [review]
LGTM -- thanks @liukun4515

File: datafusion/optimizer/src/type_coercion.rs
Comment (liukun4515) [reply datafusion/optimizer/src/type_coercion.rs:166]: I think you can got the reason from 
https://github.com/apache/arrow-datafusion/pull/3301
https://github.com/apache/arrow-datafusion/pull/3246
The unknown will be convert the physical binary op with `DistinctFrom` operation.
cc @alamb

File: datafusion/optimizer/src/type_coercion.rs
Comment (liukun4515) [reply datafusion/optimizer/src/type_coercion.rs:139]: > maybe in the long term figuring out how to avoid having to remember to call the const evaluator each struct would be good -- as it too does a expr tree walk

agree with that

Comment (liukun4515) [reply datafusion/optimizer/src/type_coercion.rs:139]: I think the SQL level with analyse/expr/optimizer should be refactor.
We are missing some formal logic like spark SQL framework or other SQL framework.
This makes it very difficult for us to do some features or optimization.

**Comment by liukun4515** [comment]
> Thanks @liukun4515 I will review this today or tomorrow

waiting for your review.

**Comment by andygrove** [comment]
> > Thanks @liukun4515 I will review this today or tomorrow
> 
> waiting for your review.

Apologies. I have not had time. I see that @alamb has approved this so please don't let me hold this up.

**Comment by liukun4515** [comment]
> > > Thanks @liukun4515 I will review this today or tomorrow
> > 
> > 
> > waiting for your review.
> 
> Apologies. I have not had time. I see that @alamb has approved this so please don't let me hold this up.

Thanks for your reply.

You can comment it after it merged.

File: datafusion/optimizer/src/type_coercion.rs
Comment (alamb) [reply datafusion/optimizer/src/type_coercion.rs:139]: I agree the current framework for expr rewriting / analysis is reaching the limits of its existing design. Refactoring the code so it is easier to work with sounds like a good idea to me. 

I am open to suggestions -- I think the challenge in any such refactors is not only coming up with a good new design but also figuring out how to incrementally get the existing code into that new structure

**Pull request merged**
